rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -381,9 +383,39 @@ class PlayerImpl
     {
       return l->recv_timestamp > r->recv_timestamp;
     }
-  } bag_message_chronological_recv_timestamp_comparator;
+  } static bag_message_chronological_recv_timestamp_comparator;
+
+  /// Comparator for SerializedBagMessageSharedPtr to order chronologically by send_timestamp.
+  struct
+  {
+    bool operator()(
+      const rosbag2_storage::SerializedBagMessageSharedPtr & l,
+      const rosbag2_storage::SerializedBagMessageSharedPtr & r) const
+    {
+      return l->send_timestamp > r->send_timestamp;
+    }
+  } static bag_message_chronological_send_timestamp_comparator;
 };
 
+decltype(PlayerImpl::bag_message_chronological_recv_timestamp_comparator)
+PlayerImpl::bag_message_chronological_recv_timestamp_comparator;
+decltype(PlayerImpl::bag_message_chronological_send_timestamp_comparator)
+PlayerImpl::bag_message_chronological_send_timestamp_comparator;
+
+PlayerImpl::BagMessageComparator PlayerImpl::get_bag_message_comparator(const MessageOrder & order)
+{
+  switch (order) {
+    case MessageOrder::RECV_TIMESTAMP:
+      return bag_message_chronological_recv_timestamp_comparator;
+    case MessageOrder::SEND_TIMESTAMP:
+      return bag_message_chronological_send_timestamp_comparator;
+    default:
+      throw std::runtime_error(
+        "unknown MessageOrder: " +
+        std::to_string(static_cast<std::underlying_type_t<MessageOrder>>(order)));
+  }
+}
+
```

Comment:
I had to do it this way because an anonymous struct was used in #1848, but my suggestion is:

```diff
diff --git a/rosbag2_transport/src/rosbag2_transport/player.cpp b/rosbag2_transport/src/rosbag2_transport/player.cpp
index 4b60f0f0..9af98a6c 100644
--- a/rosbag2_transport/src/rosbag2_transport/player.cpp
+++ b/rosbag2_transport/src/rosbag2_transport/player.cpp
@@ -375,7 +375,7 @@ private:
   static BagMessageComparator get_bag_message_comparator(const MessageOrder & order);
 
   /// Comparator for SerializedBagMessageSharedPtr to order chronologically by recv_timestamp.
-  struct
+  struct bag_message_chronological_recv_timestamp_comparator
   {
     bool operator()(
       const rosbag2_storage::SerializedBagMessageSharedPtr & l,
@@ -383,10 +383,10 @@ private:
     {
       return l->recv_timestamp > r->recv_timestamp;
     }
-  } static bag_message_chronological_recv_timestamp_comparator;
+  };
 
   /// Comparator for SerializedBagMessageSharedPtr to order chronologically by send_timestamp.
-  struct
+  struct bag_message_chronological_send_timestamp_comparator
   {
     bool operator()(
       const rosbag2_storage::SerializedBagMessageSharedPtr & l,
@@ -394,21 +394,16 @@ private:
     {
       return l->send_timestamp > r->send_timestamp;
     }
-  } static bag_message_chronological_send_timestamp_comparator;
+  };
 };
 
-decltype(PlayerImpl::bag_message_chronological_recv_timestamp_comparator)
-PlayerImpl::bag_message_chronological_recv_timestamp_comparator;
-decltype(PlayerImpl::bag_message_chronological_send_timestamp_comparator)
-PlayerImpl::bag_message_chronological_send_timestamp_comparator;
-
 PlayerImpl::BagMessageComparator PlayerImpl::get_bag_message_comparator(const MessageOrder & order)
 {
   switch (order) {
     case MessageOrder::RECV_TIMESTAMP:
-      return bag_message_chronological_recv_timestamp_comparator;
+      return bag_message_chronological_recv_timestamp_comparator{};
     case MessageOrder::SEND_TIMESTAMP:
-      return bag_message_chronological_send_timestamp_comparator;
+      return bag_message_chronological_send_timestamp_comparator{};
     default:
       throw std::runtime_error(
         "unknown MessageOrder: " +
```

Reply:
Removing those four lines does not work as-is, because static members need to be defined. In this case, it's pretty weird, since these are anonymous structs, so there is no value to initialize it to, and I have to use `decltype()` to get their types.

If I comment those four lines out, I get linking errors:

```
Starting >>> rosbag2_transport
--- stderr: rosbag2_transport                                        
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_timing.dir/build.make:204: test_play_timing] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:428: CMakeFiles/test_play_timing.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_callbacks.dir/build.make:204: test_play_callbacks] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:456: CMakeFiles/test_play_callbacks.dir/all] Error 2
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_player_stop.dir/build.make:204: test_player_stop] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:484: CMakeFiles/test_player_stop.dir/all] Error 2
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_record.dir/build.make:204: test_record] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:708: CMakeFiles/test_record.dir/all] Error 2
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_loop.dir/build.make:204: test_play_loop] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:372: CMakeFiles/test_play_loop.dir/all] Error 2
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_topic_remap.dir/build.make:204: test_play_topic_remap] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:568: CMakeFiles/test_play_topic_remap.dir/all] Error 2
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_record_all.dir/build.make:204: test_record_all] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:736: CMakeFiles/test_record_all.dir/all] Error 2
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_seek.dir/build.make:204: test_play_seek] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:512: CMakeFiles/test_play_seek.dir/all] Error 2
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_publish_clock.dir/build.make:204: test_play_publish_clock] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:400: CMakeFiles/test_play_publish_clock.dir/all] Error 2
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_next.dir/build.make:204: test_play_next] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:624: CMakeFiles/test_play_next.dir/all] Error 2
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_duration.dir/build.make:204: test_play_duration] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:596: CMakeFiles/test_play_duration.dir/all] Error 2
gmake[2]: *** [CMakeFiles/test_play_until.dir/build.make:204: test_play_until] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:652: CMakeFiles/test_play_until.dir/all] Error 2
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play_services.dir/build.make:204: test_play_services] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:540: CMakeFiles/test_play_services.dir/all] Error 2
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_composable_player.dir/build.make:213: test_composable_player] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:344: CMakeFiles/test_composable_player.dir/all] Error 2
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_burst.dir/build.make:204: test_burst] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:680: CMakeFiles/test_burst.dir/all] Error 2
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_recv_timestamp_comparator'
/usr/bin/ld: librosbag2_transport.so: undefined reference to `rosbag2_transport::PlayerImpl::bag_message_chronological_send_timestamp_comparator'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_play.dir/build.make:204: test_play] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:316: CMakeFiles/test_play.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< rosbag2_transport [40.2s, exited with code 2]
```

I don't know how it worked for you.

If I make `PlayerImpl::get_bag_message_comparator()` and the anonymous structs non-static, then we don't need these lines. However, it made more sense to me to make them static, since the class does not need to own the comparator struct instances. Let me know if you want me to change them to be non-static.

Reply:
@christophebedard That is weird. I've tried to clean cache delete all rosbag2 folders and re-copmpile and works fine for me without those 4 decltype lines. Could you please try to push deletion changes to see if CI will pass?

Reply:
done, see a1b2b070474cc5c75c957c27fd026808280135f8